File: backend/src/main/java/com/pharmacy/medlan/service/notification/NotificationService.java
package com.pharmacy.medlan.service.notification;

import com.pharmacy.medlan.dto.request.NotificationRequest;
import com.pharmacy.medlan.enums.NotificationType;
import com.pharmacy.medlan.model.notification.Notification;

import java.util.List;

/**
 * Service interface for handling notifications
 */
public interface NotificationService {
    
    /**
     * Send a notification to a user
     */
    Notification sendNotification(NotificationRequest request);
    
    /**
     * Send notification to multiple users
     */
    List<Notification> sendBulkNotifications(NotificationRequest request, List<Long> userIds);
    
    /**
     * Send notification to all users in a branch
     */
    List<Notification> sendToBranch(Long branchId, NotificationRequest request);
    
    /**
     * Send notification to users with specific role in a branch
     */
    List<Notification> sendToRole(Long branchId, String role, NotificationRequest request);
    
    /**
     * Get all notifications for a user
     */
    List<Notification> getUserNotifications(Long userId);
    
    /**
     * Get unread notifications for a user
     */
    List<Notification> getUnreadNotifications(Long userId);
    
    /**
     * Get unread notification count
     */
    int getUnreadCount(Long userId);
    
    /**
     * Mark notification as read
     */
    void markAsRead(Long notificationId);
    
    /**
     * Mark all notifications as read for a user
     */
    void markAllAsRead(Long userId);
    
    /**
     * Delete a notification
     */
    void deleteNotification(Long notificationId);
    
    /**
     * Delete all notifications for a user
     */
    void deleteAllNotifications(Long userId);
    
    /**
     * Send system alert notification
     */
    void sendSystemAlert(String title, String message, NotificationType type, Long branchId);
    
    /**
     * Send low stock alert
     */
    void sendLowStockAlert(Long productId, Long branchId, int currentStock, int reorderLevel);
    
    /**
     * Send expiry alert
     */
    void sendExpiryAlert(Long batchId, Long branchId, int daysToExpiry);
    
    /**
     * Send payment reminder
     */
    void sendPaymentReminder(Long invoiceId, Long customerId);
    
    /**
     * Send email notification
     */
    void sendEmailNotification(String to, String subject, String body);
    
    /**
     * Send SMS notification (if configured)
     */
    void sendSmsNotification(String phoneNumber, String message);
}

File: frontend/src/services/notificationService.js
/**
 * Notification Service
 * Handles all notification-related API calls
 */

import api from "@/utils/api";
import { API_ENDPOINTS } from "@/config";

const notificationService = {
  /**
   * Get user notifications
   */
  getAll: (params = {}) => {
    return api.get(API_ENDPOINTS.NOTIFICATIONS.BASE, { params });
  },

  /**
   * Get unread notifications
   */
  getUnread: () => {
    return api.get(API_ENDPOINTS.NOTIFICATIONS.UNREAD);
  },

  /**
   * Get unread notification count
   */
  getUnreadCount: () => {
    return api.get(API_ENDPOINTS.NOTIFICATIONS.UNREAD_COUNT);
  },

  /**
   * Send notification
   */
  send: (data) => {
    return api.post(API_ENDPOINTS.NOTIFICATIONS.BASE, data);
  },

  /**
   * Send bulk notifications
   */
  sendBulk: (data) => {
    return api.post(API_ENDPOINTS.NOTIFICATIONS.SEND_BULK, data);
  },

  /**
   * Send notification to branch
   */
  sendToBranch: (branchId, data) => {
    return api.post(API_ENDPOINTS.NOTIFICATIONS.SEND_TO_BRANCH(branchId), data);
  },

  /**
   * Send notification to role
   */
  sendToRole: (data) => {
    return api.post(API_ENDPOINTS.NOTIFICATIONS.SEND_TO_ROLE, data);
  },

  /**
   * Mark notification as read
   */
  markAsRead: (id) => {
    return api.put(API_ENDPOINTS.NOTIFICATIONS.MARK_READ(id));
  },

  /**
   * Mark all notifications as read
   */
  markAllAsRead: () => {
    return api.put(API_ENDPOINTS.NOTIFICATIONS.MARK_ALL_READ);
  },

  /**
   * Delete notification
   */
  delete: (id) => {
    return api.delete(API_ENDPOINTS.NOTIFICATIONS.BY_ID(id));
  },

  /**
   * Delete all notifications
   */
  deleteAll: () => {
    return api.delete(API_ENDPOINTS.NOTIFICATIONS.DELETE_ALL);
  },
};

export default notificationService;
